src/main/java/pl/project13/maven/git/NativeGitProvider.java

```diff
@@ -443,19 +443,29 @@ public boolean runEmpty(File directory, long nativeGitTimeoutInMs, String comman
         // this only works on UNIX like system not on Windows
         // ProcessBuilder builder = new ProcessBuilder(Arrays.asList("/bin/sh", "-c", command));
         // so use the same protocol as used in the run() method
-        ProcessBuilder builder = new ProcessBuilder(command.split("\\s"));
+        final ProcessBuilder builder = new ProcessBuilder(command.split("\\s"));
         final Process proc = builder.directory(directory).start();
-        final InputStream is = proc.getInputStream();
-        final InputStream err = proc.getErrorStream();
-        BufferedReader reader = new BufferedReader(new InputStreamReader(is));
+        final SubProcessOutputGobbler subProcessOutputGobbler = new SubProcessOutputGobbler(proc.getInputStream());
```

Comment:
Mhhh....since this code is now essentially somewhat is duplicated I'm starting to wonder what we actually need to fix the problem and also somewhat get rid of the code duplication here. IMHO keeping in mind that one always needs to adjust two part of the code is just a pain. Especially now since the complexity will increase.

I think in the end of the day to fix the problem we just need to read the `proc.getInputStream()` and `proc.getErrorStream()` in a separate thread (really like what is being done here inside the MR that launches a thread for each of them).

After thinking back and forth I again started to follow a different approach (sorry)...Let me elaborate...I currently see it as problem where we have the `run` function that would like to obtain all output and the `runEmpty` function that just wants to know if there is one line of output. Besides that both functions should pretty much do the same stuff with building a process and now building a thread structure to read `proc.getInputStream()` and `proc.getErrorStream()`.

In general I now tried to factor out the entire logic of building the thread structure and essentially ended up with creating a function that accepts a string (e.g. the content we read from either `proc.getInputStream()` or `proc.getErrorStream()`) and that function can return a boolean to indicate if it wants to continue reading the stream, or doesn't. The signature boils down to ``Function<String, Boolean>`` and converts the both read functions into the following:
```java
@Override
    public String run(File directory, long nativeGitTimeoutInMs, String command) throws IOException {
      String output = "";
      try {
        final StringBuilder commandResult = new StringBuilder();

        final Function<String, Boolean> stdoutConsumer = line -> {
          if (line != null) {
            commandResult.append(line).append("\n");
          }
          // return true to indicate we want to read more content
          return true;
        };
        runProcess(directory, nativeGitTimeoutInMs, command, stdoutConsumer);

        output = commandResult.toString();
      } catch (final InterruptedException ex) {
        throw new IOException(ex);
      }
      return output;
    }

    @Override
    public boolean runEmpty(File directory, long nativeGitTimeoutInMs, String command) throws IOException {
      final AtomicBoolean empty = new AtomicBoolean(true);

      try {
        final Function<String, Boolean> stdoutConsumer = line -> {
          if (line != null) {
            empty.set(false);
          }
          // return false to indicate we don't need to read more content
          return false;
        };
        runProcess(directory, nativeGitTimeoutInMs, command, stdoutConsumer);
      } catch (final InterruptedException ex) {
        throw new IOException(ex);
      }
      return empty.get(); // was non-empty
    }
```

All the we "build a thread structure" and "process error handling" is residing within `runProcess` and thus providing the same code to both functions.
See [here for for full reference](https://github.com/git-commit-id/maven-git-commit-id-plugin/files/2602607/396.txt).

WDYT? Does this look reasonable? Would that fix the problem? (Again, on my side everything seems to be working...)

Reply:
As far as a i can see your modifications seem reasonable. So feel free to continue and use my PR as base line for your refactorings, which really look nice 👍

Reply:
Thanks for your feedback and especially to your contribution on this project!
Will do that :-)